### ISSUE TITLE
Clean up arrays memory handling flags

### ISSUE THREAD

**Issue by EmilyBourne** [pyccel/pyccel#1085]
There are 3 types of arrays:
- pointers
- stack arrays
- allocatables

These 3 options are mutually exclusive. E.g. an object cannot be both a pointer and allocatable. One of these 3 options must always be set to `True` (otherwise the array cannot be declared except in the case of an argument, which can be treated like a stack array).

This is obvious to developers familiar with Fortran, however from a C perspective it is less clear. In order to avoid bugs I propose that we enforce these rules by storing a string in an attribute `Variable._memory_handling` which can only have one of the following values: `pointer`, `stack_array`, `allocatable`.

This issue was created following confusion in PR [#1056](https://github.com/pyccel/pyccel/pull/1056#discussion_r820478765)

**Comment by EmilyBourne** [comment]
It may also be worth renaming `allocatable`->`heap_array` for clarity. What do you think @yguclu @saidctb ?

**Comment by yguclu** [comment]
> It may also be worth renaming `allocatable`->`heap_array` for clarity.

I wonder if we want to use a naming scheme that is more general, so that it can be applied to class objects too. For example we could use `alias` (or `reference`), `stack`, and `heap`.

**Comment by EmilyBourne** [comment]
> I wonder if we want to use a naming scheme that is more general, so that it can be applied to class objects too. For example we could use `alias` (or `reference`), `stack`, and `heap`.

Anything more general sounds good to me, but I'm not entirely sure what you mean by `alias`. Is this just to avoid using `pointer` or `reference` to avoid potential future connotations in languages that have both these concepts?

**Comment by yguclu** [comment]
> > I wonder if we want to use a naming scheme that is more general, so that it can be applied to class objects too. For example we could use `alias` (or `reference`), `stack`, and `heap`.
> 
> Anything more general sounds good to me, but I'm not entirely sure what you mean by `alias`. Is this just to avoid using `pointer` or `reference` to avoid potential future connotations in languages that have both these concepts?

Yes and no. By `alias` I mean that in the Python code we are using another variable to refer to an object that already exists. So one can see this as just another name for the object, hence an "alias". As an additional benefit, we isolate the semantic parser from an implementation detail (i.e. using a `pointer` or a `reference`) which depends on the target language, and should therefore belong to the printer.

**Comment by aihya** [comment]
> In order to avoid bugs I propose that we enforce these rules by storing a string in an attribute Variable._memory_handling which can only have one of the following values: pointer, stack_array, allocatable.

Does this mean that the three attributes `Variable.allocatable`, `Variable.is_pointer` and `Variable.is_stack_array` must be replaced with `Variable._memory_handling` and remove all occurrences of the first three, or just add as an extra attribute?

**Comment by EmilyBourne** [comment]
> > In order to avoid bugs I propose that we enforce these rules by storing a string in an attribute Variable._memory_handling which can only have one of the following values: pointer, stack_array, allocatable.
> 
> Does this mean that the three attributes `Variable.allocatable`, `Variable.is_pointer` and `Variable.is_stack_array` must be replaced with `Variable._memory_handling` and remove all occurrences of the first three, or just add `Variable._memory_handling` as an extra attribute?

As you say, the three should be replaced. I think having 4 variables would be error prone. They are likely to become out of sync. Properties can be used to keep a similar interface though

**Comment by EmilyBourne** [comment]
Historically we have always used `Variable.allocatable`, `Variable.is_stack_array` and `Variable.is_pointer` for arrays. This is why sometimes all 3 are set to `False`. I think we have 2 options here:

1. Use `None` for scalars
2. Use `stack` for scalars

Personally I prefer the second option. It more accurately reflects how the memory is handled and if we are using `stack`, `heap` and `alias` then the names are not misleading

**Comment by EmilyBourne** [comment]
For your 3 examples:

- https://github.com/pyccel/pyccel/blob/fdc67a8cf005b92aba53453ec5d47a32ac5b29a8/pyccel/parser/semantic.py#L414
    This currently sets the allocatable (heap) property if the object is an array (otherwise it is a scalar allocated on the stack which currently doesn't need a property to be set).
     After fixing this issue, all that would need to be done is to pass along the `memory_handling` property
- https://github.com/pyccel/pyccel/blob/fdc67a8cf005b92aba53453ec5d47a32ac5b29a8/pyccel/parser/semantic.py#L443
    The `Concatenate` operator is used for concatenating tuples (`(1,2)+(3,4,True)`). Here we set the array to a heap array if it contains a heap array (i.e. an unknown size), otherwise it is treated as a stack array
- https://github.com/pyccel/pyccel/blob/fdc67a8cf005b92aba53453ec5d47a32ac5b29a8/pyccel/parser/semantic.py#L465
    Arrays are created on the heap by default. Here we simply check that we are dealing with an array (otherwise we can allocate on the stack for scalars)
